src/Parser/HolyPriest/Modules/PriestCore/Serendipity.js

```diff
@@ -0,0 +1,227 @@
+
+const HOLY_WORD_SPELL_IDS = [
+  SPELLS.HOLY_WORD_SERENITY.id,
+  SPELLS.HOLY_WORD_SANCTIFY.id,
+];
+
+// This module might contain too much information but to rewrite some of the output
+// in this file would entail a lot of redundant copying (aka violating the "DRY"
+// coding mindset). Let me @enragednuke if you think I should move stuff out.
+
+// Currently contains:
+// - "Wasted Serendipity" statistic
+// - "Fully missed holy word casts (via Serendipity)" suggestion
+// - "Holy Priest T20 2P Bonus" item
+class Serendipity extends Module {
+  static dependencies = {
+    combatants: Combatants,
+  }
+
+  // Holy Word reduction spells (aka things that apply Serendipity)
+  serendipityProccers = {
+    [SPELLS.GREATER_HEAL.id]: SPELLS.HOLY_WORD_SERENITY.id,
+    [SPELLS.FLASH_HEAL.id]: SPELLS.HOLY_WORD_SERENITY.id,
+    [SPELLS.PRAYER_OF_HEALING.id]: SPELLS.HOLY_WORD_SANCTIFY.id,
+  }
+
+  getAffectedSpell(spellId) {
+    if (Object.keys(this.serendipityProccers).indexOf(spellId.toString()) === -1) {
+      return -1;
+    }
+    return this.serendipityProccers[spellId];
+  }
+
+  // CD tracker
+  holyWordCooldowns = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 0,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0,
+  }
+
+  // Cooldowns for Holy Words
+  holyWordMaxCooldown = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 60,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0, // modified in on_initialized
+  }
+
+  // Serendipity amount (modified accordingly in on_initialized)
+  serendipityReduction = 6.0
+
+
+  //   Output values
+
+  holy2PsetValue = 0.0
+
+  // Overall wasted serendipity
+  overcast = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 0.0,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0.0,
+  }
+  // Cast tracker
+  casts = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 0,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0,
+  }
+  // How many full casts were missed (i.e. 60s of consecutive wasted Serendipity on Serenity)
+  fullOvercast = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 0,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0,
+  }
+  // Tracker of how much wasted overcast between each holy word cast
+  _tempOvercast = {
+    [SPELLS.HOLY_WORD_SERENITY.id]: 0.0,
+    [SPELLS.HOLY_WORD_SANCTIFY.id]: 0.0,
+  }
+
+  on_initialized() {
+    // Set up proper serendipity reduction values
+    if (this.combatants.selected.hasTalent(SPELLS.LIGHT_OF_THE_NAARU_TALENT.id)) {
+      this.serendipityReduction += 2.0;
+    }
+    if (this.combatants.selected.hasBuff(SPELLS.HOLY_PRIEST_T20_2SET_BONUS_BUFF)) {
+      this.serendipityReduction += 1.0;
+    }
+
+    // Check for Piety for serendipityProccers
+    if (this.combatants.selected.hasTalent(SPELLS.PIETY_TALENT.id)) {
+      this.serendipityProccers[SPELLS.PRAYER_OF_MENDING_CAST.id] = SPELLS.HOLY_WORD_SANCTIFY.id;
+    }
+
+    // Modify Sanctify CD based on trait
+    this.holyWordMaxCooldown[SPELLS.HOLY_WORD_SANCTIFY.id] = 60 - ((this.combatants.selected.traitsBySpellId[SPELLS.HALLOWED_GROUND_TRAIT.id] || 0) * SPELLS.HALLOWED_GROUND_TRAIT.coeff);
```

Comment:
The other talented/traited things are given a default value which is then modified on_initialized(), like with Light of the Naaru above. Why are you handling this differently?

Reply:
Ah right, sorry. My main language isn't JavaScript so I forget calling for an undefined index won't throw errors.